### ISSUE TITLE
Surprising behavior of shared axes with categorical units

### ISSUE THREAD

**Issue by mwaskom** [matplotlib/matplotlib#18273]
### Bug report

Setting categorical units on an axis that is shared across subplots updates the units on its siblings, but not the order of the units.

**Code for reproduction**

```python
f, axs = plt.subplots(1, 2, sharex=True)
axs[0].xaxis.update_units(["a", "c", "b"])
axs[1].plot(["a", "b", "c"], [1, 2, 3])
```

**Actual outcome**

![image](https://user-images.githubusercontent.com/315810/90342440-63e26c00-dfd6-11ea-8661-57b8cc37c47c.png)

**Expected outcome**

Compare to:

```python
f, axs = plt.subplots(1, 2, sharex=True)
axs[0].xaxis.update_units(["a", "c", "b"])
axs[0].plot(["a", "b", "c"], [1, 2, 3])
```

![image](https://user-images.githubusercontent.com/315810/90342454-752b7880-dfd6-11ea-8fa1-b406e1c91699.png)

**Matplotlib version**

  * Operating system:
  * Matplotlib version: 3.3.0
  * Matplotlib backend (`print(matplotlib.get_backend())`): pylab inline
  * Python version: 3.8

**Comment by aitikgupta** [comment]
I think this has been resolved, since the issue is no longer reproducible on `master` branch.
Although, it is still reproducible for `matplotlib==3.3.0`

**Comment by QuLogic** [comment]
Yes, this was fixed by #18308.
